File: Source/UI/Multiboard/StandardMultiboard.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;

public static class StandardMultiboard
{
    public static multiboard OverallStats;
    public static multiboard CurrentStats;
    public static multiboard BestTimes;
    private static trigger Updater;
    private static trigger ESCTrigger;

    private static string color = Colors.COLOR_YELLOW_ORANGE;
    private static string[] PlayerStats = new string[8];
    private static float[] RoundTimes = new float[5];
    private static List<player> PlayersList = new List<player>();

    public static void Initialize()
    {
        if (Gamemode.CurrentGameMode != GameMode.Standard) return;
        BestTimes = multiboard.Create();
        OverallStats = multiboard.Create();
        CurrentStats = multiboard.Create();
        Init();
    }

    /// <summary>
    /// Wait till difficulty is chosen, then begin..
    /// </summary>
    private static void Init()
    {
        var t = ObjectPool<AchesTimers>.GetEmptyObject();
        t.Timer.Start(1.0f, true, ErrorHandler.Wrap(() =>
        {
            if (!Difficulty.IsDifficultyChosen) return;
            MakeMultiboard();
            RegisterTriggers();
            t.Dispose();
        }));
    }

    private static void RegisterTriggers()
    {
        Updater = trigger.Create();
        ESCTrigger = trigger.Create();

        Updater.RegisterTimerEvent(1.00f, true);
        Updater.AddAction(CurrentStatsRoundTimes);

        foreach (var player in Globals.ALL_PLAYERS)
            ESCTrigger.RegisterPlayerEvent(player, playerevent.EndCinematic);
        ESCTrigger.AddAction(ESCPressed);
    }

    private static void MakeMultiboard()
    {
        OverallGamesStatsMultiboard();
        BestTimesMultiboard();
        CurrentGameStatsMultiboard();
        CurrentGameStats();
    }

    private static void CurrentGameStatsMultiboard()
    {
        CurrentStats.Rows = Globals.ALL_PLAYERS.Count + 2;
        CurrentStats.Columns = 7;
        CurrentStats.GetItem(0, 0).SetText($"{color}Round Time|r");
        CurrentStats.GetItem(1, 0).SetText($"{color}Player|r");
        CurrentStats.GetItem(1, 1).SetText($"{color}Score|r");
        CurrentStats.GetItem(1, 2).SetText($"{color}Saves|r");
        CurrentStats.GetItem(1, 3).SetText($"{color}Deaths|r");
        CurrentStats.GetItem(1, 4).SetText($"{color}Streak|r");
        CurrentStats.GetItem(1, 5).SetText($"{color}Ratio|r");
        CurrentStats.GetItem(1, 6).SetText($"{color}S / D|r");
        CurrentStats.SetChildVisibility(true, false);
        CurrentStats.SetChildWidth(0.055f);
        CurrentStats.GetItem(1, 0).SetWidth(0.07f);
        CurrentStats.IsDisplayed = true;
    }

    private static void OverallGamesStatsMultiboard()
    {
        OverallStats.Rows = Globals.ALL_PLAYERS.Count + 1;
        OverallStats.Columns = 8;
        OverallStats.GetItem(0, 0).SetText($"{color}Player|r");
        OverallStats.GetItem(0, 1).SetText($"{color}Score:|r");
        OverallStats.GetItem(0, 2).SetText($"{color}Saves|r");
        OverallStats.GetItem(0, 3).SetText($"{color}Deaths|r");
        OverallStats.GetItem(0, 4).SetText($"{color}MaxStreak|r");
        OverallStats.GetItem(0, 5).SetText($"{color}  Ratio|r");
        OverallStats.GetItem(0, 6).SetText($"{color}Games|r");
        OverallStats.GetItem(0, 7).SetText($"{color}Wins|r");
        OverallStats.SetChildVisibility(true, false);
        OverallStats.SetChildWidth(0.052f);
        OverallStats.GetItem(0, 0).SetWidth(0.07f);
        OverallStats.IsDisplayed = false;
        UpdateOverallStatsMB();
    }

    private static void BestTimesMultiboard()
    {
        BestTimes.Rows = Globals.ALL_PLAYERS.Count + 1;
        BestTimes.Columns = 6;
        BestTimes.GetItem(0, 0).SetText($"{color}Player|r");
        BestTimes.GetItem(0, 1).SetText($"{color}Round 1|r");
        BestTimes.GetItem(0, 2).SetText($"{color}Round 2|r");
        BestTimes.GetItem(0, 3).SetText($"{color}Round 3|r");
        BestTimes.GetItem(0, 4).SetText($"{color}Round 4|r");
        BestTimes.GetItem(0, 5).SetText($"{color}Round 5|r");
        BestTimes.SetChildVisibility(true, false);
        BestTimes.SetChildWidth(0.05f);
        BestTimes.GetItem(0, 0).SetWidth(0.07f);
        BestTimes.IsDisplayed = false;
        UpdateBestTimesMB();
    }

    private static void CurrentStatsRoundTimes()
    {
        var roundColor = Colors.COLOR_GREEN;
        var totalColor = Colors.COLOR_YELLOW_ORANGE;
        var totalTime = 0.0f;
        for (int i = 1; i <= Gamemode.NumberOfRounds; i++)
        {
            CurrentStats.GetItem(0, i).SetText($"{roundColor}{Utility.ConvertFloatToTimeInt(GameTimer.FinishedTimes[i])}|r");
            totalTime += GameTimer.FinishedTimes[i];
        }
        CurrentStats.GetItem(0, Gamemode.NumberOfRounds + 1).SetText($"{totalColor}{Utility.ConvertFloatToTimeInt(totalTime)}|r");

    }

    private static void CurrentGameStats()
    {
        try
        {
            CurrentStats.Title = $"Current Stats {Colors.COLOR_YELLOW_ORANGE}[{Gamemode.CurrentGameMode}-{Difficulty.DifficultyOption.ToString()}]|r {Colors.COLOR_RED}[Press ESC]|r";
            CurrentStats.Rows = Globals.ALL_PLAYERS.Count + 2;
            var rowIndex = 2;

            // Use a list to hold keys for manual sorting
            PlayersList.Clear();

            for(int i = 0; i < Globals.ALL_PLAYERS.Count; i++)
            {
                PlayersList.Add(Globals.ALL_PLAYERS[i]);
            }

            // Sort the array of keys based on custom criteria
            for (int i = 0; i < PlayersList.Count; i++)
            {
                for (int j = i + 1; j < PlayersList.Count; j++)
                {
                    var stats1 = Globals.ALL_KITTIES[PlayersList[i]].CurrentStats;
                    var stats2 = Globals.ALL_KITTIES[PlayersList[j]].CurrentStats;
                    var score1 = stats1.TotalSaves - stats1.TotalDeaths;
                    var score2 = stats2.TotalSaves - stats2.TotalDeaths;

                    if (score2 > score1 || (score2 == score1 && PlayersList[j].Id < PlayersList[i].Id))
                    {
                        var temp = PlayersList[i];
                        PlayersList[i] = PlayersList[j];
                        PlayersList[j] = temp;
                    }
                }
            }

            for (int i = 0; i < PlayersList.Count; i++)
            {
                var player = PlayersList[i];
                var currentStats = Globals.ALL_KITTIES[player].CurrentStats;
                var playerColor = Colors.GetStringColorOfPlayer(player.Id + 1);

                var name = player.Name.Length > 8 ? player.Name.Substring(0, 8) : player.Name;
                var score = currentStats.TotalSaves - currentStats.TotalDeaths;
                var kda = currentStats.TotalDeaths == 0
                    ? currentStats.TotalSaves.ToString("F2")
                    : (currentStats.TotalSaves / (double)currentStats.TotalDeaths).ToString("F2");

                PlayerStats[0] = name;
                PlayerStats[1] = score.ToString();
                PlayerStats[2] = currentStats.TotalSaves.ToString();
                PlayerStats[3] = currentStats.TotalDeaths.ToString();
                PlayerStats[4] = currentStats.SaveStreak.ToString();
                PlayerStats[5] = kda;
                PlayerStats[6] = $"{currentStats.RoundSaves} / {currentStats.RoundDeaths}";

                for (int j = 0; j < PlayerStats.Length - 1; j++) // skip last element
                {
                    CurrentStats.GetItem(rowIndex, j).SetText($"{playerColor}{PlayerStats[j]}{Colors.COLOR_RESET}");
                    if (j == 0) CurrentStats.GetItem(rowIndex, j).SetWidth(0.07f);
                }

                rowIndex++;
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"{Colors.COLOR_DARK_RED}Error in CurrentGameStats multiboard: {ex.Message}");
        }
    }

    private static void OverallGameStats()
    {
        OverallStats.Title = $"Overall Stats {Colors.COLOR_YELLOW_ORANGE}[{Gamemode.CurrentGameMode}-{Difficulty.DifficultyOption.ToString()}]|r {Colors.COLOR_RED}[Press ESC]|r";
        OverallStats.Rows = Globals.ALL_PLAYERS.Count + 1;
        var rowIndex = 1;

        PlayersList.Clear();
        for (int i = 0; i < Globals.ALL_PLAYERS.Count; i++)
        {
            PlayersList.Add(Globals.ALL_PLAYERS[i]);
        }

        // Sort the array of keys based on custom criteria
        for (int i = 0; i < PlayersList.Count; i++)
        {
            for (int j = i + 1; j < PlayersList.Count; j++)
            {
                var stats1 = Globals.ALL_KITTIES[PlayersList[i]].SaveData.GameStats;
                var stats2 = Globals.ALL_KITTIES[PlayersList[j]].SaveData.GameStats;
                var score1 = stats1.Saves - stats1.Deaths;
                var score2 = stats2.Saves - stats2.Deaths;

                if (score2 > score1 || (score2 == score1 && PlayersList[j].Id < PlayersList[i].Id))
                {
                    var temp = PlayersList[i];
                    PlayersList[i] = PlayersList[j];
                    PlayersList[j] = temp;
                }
            }
        }

        for(int i = 0; i < PlayersList.Count; i++)
        {
            var player = PlayersList[i];
            var saveData = Globals.ALL_KITTIES[player].SaveData;
            var playerColor = Colors.GetStringColorOfPlayer(player.Id + 1);

            var name = player.Name.Length > 8 ? player.Name.Substring(0, 8) : player.Name;
            var allSaves = saveData.GameStats.Saves;
            var allDeaths = saveData.GameStats.Deaths;
            var score = allSaves - allDeaths;
            var kda = allDeaths == 0 ? allSaves.ToString("F2") : (allSaves / (double)allDeaths).ToString("F2");
            var (games, wins) = GetGameStatData(saveData);

            PlayerStats[0] = name;
            PlayerStats[1] = score.ToString();
            PlayerStats[2] = allSaves.ToString();
            PlayerStats[3] = allDeaths.ToString();
            PlayerStats[4] = saveData.GameStats.HighestSaveStreak.ToString();
            PlayerStats[5] = kda;
            PlayerStats[6] = games.ToString();
            PlayerStats[7] = wins.ToString();

            for (int j = 0; j < PlayerStats.Length; j++)
            {
                OverallStats.GetItem(rowIndex, j).SetText($"{playerColor}{PlayerStats[j]}{Colors.COLOR_RESET}");
                if (j == 0) OverallStats.GetItem(rowIndex, j).SetWidth(0.07f);
            }

            rowIndex++;
        }
    }

    private static void BestTimesStats()
    {
        BestTimes.Title = $"Best Times {Colors.COLOR_YELLOW_ORANGE}[{Gamemode.CurrentGameMode}-{Difficulty.DifficultyOption.ToString()}]|r {Colors.COLOR_RED}[Press ESC]|r";
        BestTimes.Rows = Globals.ALL_PLAYERS.Count + 1;
        var rowIndex = 1;

        for(int i = 0; i < Globals.ALL_PLAYERS.Count; i++)
        {
            var player = Globals.ALL_PLAYERS[i];
            var saveData = Globals.ALL_KITTIES[player].SaveData;
            var playerColor = Colors.GetStringColorOfPlayer(player.Id + 1);

            var roundTimes = GetGameRoundTime(saveData);

            for (int j = 0; j < roundTimes.Length; j++)
            {
                if (roundTimes[j] != 0)
                    BestTimes.GetItem(rowIndex, j + 1).SetText($"{playerColor}{Utility.ConvertFloatToTime(roundTimes[j])}{Colors.COLOR_RESET}");
                else
                    BestTimes.GetItem(rowIndex, j + 1).SetText($"{playerColor}---{Colors.COLOR_RESET}");
            }
            rowIndex++;
        }
    }

    public static void UpdateOverallStatsMB()
    {
        if (Gamemode.CurrentGameMode != GameMode.Standard) return;
        OverallGameStats();
    }

    public static void UpdateStandardCurrentStatsMB()
    {
        if (Gamemode.CurrentGameMode != GameMode.Standard) return;
        CurrentGameStats();
    }

    public static void UpdateBestTimesMB()
    {
        if (Gamemode.CurrentGameMode != GameMode.Standard) return;
        MultiboardUtil.FillPlayers(BestTimes, 1);
        BestTimesStats();
    }

    private static (int gameCount, int winCount) GetGameStatData(KittyData data)
    {
        var gameData = data.GameStats;
        int numberOfGames;
        int numberOfWins;
        switch (Difficulty.DifficultyValue)
        {
            case (int)DifficultyLevel.Normal:
                numberOfGames = gameData.NormalGames;
                numberOfWins = gameData.NormalWins;
                break;

            case (int)DifficultyLevel.Hard:
                numberOfGames = gameData.HardGames;
                numberOfWins = gameData.HardWins;
                break;

            case (int)DifficultyLevel.Impossible:
                numberOfGames = gameData.ImpossibleGames;
                numberOfWins = gameData.ImpossibleWins;
                break;
            case (int)DifficultyLevel.Nightmare:
                numberOfGames = gameData.NightmareGames;
                numberOfWins = gameData.NightmareWins;
                break;
            default:
                Console.WriteLine($"{Colors.COLOR_DARK_RED}Error multiboard getting gamestat data.");
                return (0, 0);
        }
        return (numberOfGames, numberOfWins);
    }

    private static float[] GetGameRoundTime(KittyData data)
    {
        var gameData = data.RoundTimes;

        switch (Difficulty.DifficultyValue)
        {
            case (int)DifficultyLevel.Normal:
                RoundTimes[0] = gameData.RoundOneNormal;
                RoundTimes[1] = gameData.RoundTwoNormal;
                RoundTimes[2] = gameData.RoundThreeNormal;
                RoundTimes[3] = gameData.RoundFourNormal;
                RoundTimes[4] = gameData.RoundFiveNormal;
                break;

            case (int)DifficultyLevel.Hard:
                RoundTimes[0] = gameData.RoundOneHard;
                RoundTimes[1] = gameData.RoundTwoHard;
                RoundTimes[2] = gameData.RoundThreeHard;
                RoundTimes[3] = gameData.RoundFourHard;
                RoundTimes[4] = gameData.RoundFiveHard;
                break;

            case (int)DifficultyLevel.Impossible:
                RoundTimes[0] = gameData.RoundOneImpossible;
                RoundTimes[1] = gameData.RoundTwoImpossible;
                RoundTimes[2] = gameData.RoundThreeImpossible;
                RoundTimes[3] = gameData.RoundFourImpossible;
                RoundTimes[4] = gameData.RoundFiveImpossible;
                break;
            case (int)DifficultyLevel.Nightmare:
                RoundTimes[0] = gameData.RoundOneNightmare;
                RoundTimes[1] = gameData.RoundTwoNightmare;
                RoundTimes[2] = gameData.RoundThreeNightmare;
                RoundTimes[3] = gameData.RoundFourNightmare;
                RoundTimes[4] = gameData.RoundFiveNightmare;
                break;
            default:
                Console.WriteLine($"{Colors.COLOR_DARK_RED}Error multiboard getting gamestat data.");
                return RoundTimes;
        }
        return RoundTimes;
    }

    private static void ESCPressed()
    {
        if (Gamemode.CurrentGameMode != GameMode.Standard) return;
        if (!@event.Player.IsLocal) return;
        if (CurrentStats.IsDisplayed)
        {
            CurrentStats.IsDisplayed = false;
            OverallStats.IsDisplayed = true;
        }
        else if (OverallStats.IsDisplayed)
        {
            OverallStats.IsDisplayed = false;
            BestTimes.IsDisplayed = true;
        }
        else if (BestTimes.IsDisplayed)
        {
            BestTimes.IsDisplayed = false;
            CurrentStats.IsDisplayed = true;
        }
    }
}

File: src/UI/Multiboard/StandardMultiboard.ts
import { GameTimer } from 'src/Game/Rounds/GameTimer'
import { CurrentGameMode } from 'src/Gamemodes/CurrentGameMode'
import { GameMode } from 'src/Gamemodes/GameModeEnum'
import { Globals } from 'src/Global/Globals'
import { Difficulty } from 'src/Init/Difficulty/Difficulty'
import { DifficultyLevel } from 'src/Init/Difficulty/DifficultyOption'
import { KittyData } from 'src/SaveSystem2.0/MAKE REWARDS HERE/KittyData'
import { Colors } from 'src/Utility/Colors/Colors'
import { ColorUtils } from 'src/Utility/Colors/ColorUtils'
import { Utility } from 'src/Utility/Utility'
import { getTriggerPlayer } from 'src/Utility/w3tsUtils'
import { MapPlayer, Multiboard, Timer, Trigger } from 'w3ts'

export class StandardMultiboard {
    public static OverallStats: Multiboard
    public static CurrentStats: Multiboard
    public static BestTimes: Multiboard
    private static Updater: Trigger
    private static ESCTrigger: Trigger

    private static color: string = Colors.COLOR_YELLOW_ORANGE
    private static PlayerStats: string[] = []
    private static RoundTimes: number[] = []
    private static PlayersList: MapPlayer[] = []

    public static Initialize = () => {
        if (CurrentGameMode.active !== GameMode.Standard) return
        StandardMultiboard.BestTimes = Multiboard.create()!
        StandardMultiboard.OverallStats = Multiboard.create()!
        StandardMultiboard.CurrentStats = Multiboard.create()!
        StandardMultiboard.Init()
    }

    /// <summary>
    /// Wait till difficulty is chosen, then begin..
    /// </summary>
    private static Init = () => {
        const t = Timer.create()
        t.start(1.0, true, () => {
            if (!Difficulty.IsDifficultyChosen) return
            StandardMultiboard.MakeMultiboard()
            StandardMultiboard.RegisterTriggers()
            t.destroy()
        })
    }

    private static RegisterTriggers = () => {
        StandardMultiboard.Updater = Trigger.create()!
        StandardMultiboard.ESCTrigger = Trigger.create()!

        StandardMultiboard.Updater.registerTimerEvent(1.0, true)
        StandardMultiboard.Updater.addAction(() => StandardMultiboard.CurrentStatsRoundTimes())

        for (const player of Globals.ALL_PLAYERS)
            StandardMultiboard.ESCTrigger.registerPlayerEvent(player, EVENT_PLAYER_END_CINEMATIC)
        StandardMultiboard.ESCTrigger.addAction(() => StandardMultiboard.ESCPressed())
    }

    private static MakeMultiboard = () => {
        print('Making Standard Multiboard')
        StandardMultiboard.OverallGamesStatsMultiboard()
        StandardMultiboard.BestTimesMultiboard()
        StandardMultiboard.CurrentGameStatsMultiboard()
        StandardMultiboard.CurrentGameStats()
    }

    private static CurrentGameStatsMultiboard = () => {
        StandardMultiboard.CurrentStats.rows = Globals.ALL_PLAYERS.length + 2
        StandardMultiboard.CurrentStats.columns = 7
        StandardMultiboard.CurrentStats.GetItem(0, 0).setText(`${StandardMultiboard.color}Time: Round|r`)
        StandardMultiboard.CurrentStats.GetItem(1, 0).setText(`${StandardMultiboard.color}Player|r`)
        StandardMultiboard.CurrentStats.GetItem(1, 1).setText(`${StandardMultiboard.color}Score|r`)
        StandardMultiboard.CurrentStats.GetItem(1, 2).setText(`${StandardMultiboard.color}Saves|r`)
        StandardMultiboard.CurrentStats.GetItem(1, 3).setText(`${StandardMultiboard.color}Deaths|r`)
        StandardMultiboard.CurrentStats.GetItem(1, 4).setText(`${StandardMultiboard.color}Streak|r`)
        StandardMultiboard.CurrentStats.GetItem(1, 5).setText(`${StandardMultiboard.color}Ratio|r`)
        StandardMultiboard.CurrentStats.GetItem(1, 6).setText(`${StandardMultiboard.color}S / D|r`)
        StandardMultiboard.CurrentStats.SetChildVisibility(true, false)
        StandardMultiboard.CurrentStats.setItemsWidth(0.055)
        StandardMultiboard.CurrentStats.GetItem(1, 0).setWidth(0.07)
        StandardMultiboard.CurrentStats.display(true)
    }

    private static OverallGamesStatsMultiboard = () => {
        StandardMultiboard.OverallStats.rows = Globals.ALL_PLAYERS.length + 1
        StandardMultiboard.OverallStats.columns = 8
        StandardMultiboard.OverallStats.GetItem(0, 0).setText(`${StandardMultiboard.color}Player|r`)
        StandardMultiboard.OverallStats.GetItem(0, 1).setText(`${StandardMultiboard.color}Score:|r`)
        StandardMultiboard.OverallStats.GetItem(0, 2).setText(`${StandardMultiboard.color}Saves|r`)
        StandardMultiboard.OverallStats.GetItem(0, 3).setText(`${StandardMultiboard.color}Deaths|r`)
        StandardMultiboard.OverallStats.GetItem(0, 4).setText(`${StandardMultiboard.color}MaxStreak|r`)
        StandardMultiboard.OverallStats.GetItem(0, 5).setText(`${StandardMultiboard.color}  Ratio|r`)
        StandardMultiboard.OverallStats.GetItem(0, 6).setText(`${StandardMultiboard.color}Games|r`)
        StandardMultiboard.OverallStats.GetItem(0, 7).setText(`${StandardMultiboard.color}Wins|r`)
        StandardMultiboard.OverallStats.SetChildVisibility(true, false)
        StandardMultiboard.OverallStats.setItemsWidth(0.052)
        StandardMultiboard.OverallStats.GetItem(0, 0).setWidth(0.07)
        StandardMultiboard.OverallStats.display(false)
        StandardMultiboard.UpdateOverallStatsMB()
    }

    private static BestTimesMultiboard = () => {
        StandardMultiboard.BestTimes.rows = Globals.ALL_PLAYERS.length + 1
        StandardMultiboard.BestTimes.columns = 6
        StandardMultiboard.BestTimes.GetItem(0, 0).setText(`${StandardMultiboard.color}Player|r`)
        StandardMultiboard.BestTimes.GetItem(0, 1).setText(`${StandardMultiboard.color}1: Round|r`)
        StandardMultiboard.BestTimes.GetItem(0, 2).setText(`${StandardMultiboard.color}2: Round|r`)
        StandardMultiboard.BestTimes.GetItem(0, 3).setText(`${StandardMultiboard.color}3: Round|r`)
        StandardMultiboard.BestTimes.GetItem(0, 4).setText(`${StandardMultiboard.color}4: Round|r`)
        StandardMultiboard.BestTimes.GetItem(0, 5).setText(`${StandardMultiboard.color}5: Round|r`)
        StandardMultiboard.BestTimes.SetChildVisibility(true, false)
        StandardMultiboard.BestTimes.setItemsWidth(0.05)
        StandardMultiboard.BestTimes.GetItem(0, 0).setWidth(0.07)
        StandardMultiboard.BestTimes.display(false)
        StandardMultiboard.UpdateBestTimesMB()
    }

    private static CurrentStatsRoundTimes = () => {
        const color = Colors.COLOR_GREEN
        for (let i = 1; i <= Globals.NumberOfRounds; i++) {
            StandardMultiboard.CurrentStats.GetItem(0, i).setText(
                `${color}${Utility.ConvertFloatToTimeInt(GameTimer.RoundTime[i] || 0)}|r`
            )
        }
    }

    private static CurrentGameStats = () => {
        try {
            StandardMultiboard.CurrentStats.title = `${Colors.COLOR_YELLOW_ORANGE}Current Stats [${CurrentGameMode.active}-${Difficulty.DifficultyOption.toString()}]|r ${Colors.COLOR_RED}[Press ESC]|r`
            StandardMultiboard.CurrentStats.rows = Globals.ALL_PLAYERS.length + 2
            let rowIndex = 2

            // Use a list to hold keys for manual sorting
            StandardMultiboard.PlayersList.length = 0

            for (let i = 0; i < Globals.ALL_PLAYERS.length; i++) {
                StandardMultiboard.PlayersList.push(Globals.ALL_PLAYERS[i])
            }

            // Sort the array of keys based on custom criteria
            for (let i = 0; i < StandardMultiboard.PlayersList.length; i++) {
                for (let j: number = i + 1; j < StandardMultiboard.PlayersList.length; j++) {
                    const stats1 = Globals.ALL_KITTIES.get(StandardMultiboard.PlayersList[i])!.CurrentStats
                    const stats2 = Globals.ALL_KITTIES.get(StandardMultiboard.PlayersList[j])!.CurrentStats
                    const score1 = stats1.TotalSaves - stats1.TotalDeaths
                    const score2 = stats2.TotalSaves - stats2.TotalDeaths

                    if (
                        score2 > score1 ||
                        (score2 === score1 &&
                            StandardMultiboard.PlayersList[j].id < StandardMultiboard.PlayersList[i].id)
                    ) {
                        const temp = StandardMultiboard.PlayersList[i]
                        StandardMultiboard.PlayersList[i] = StandardMultiboard.PlayersList[j]
                        StandardMultiboard.PlayersList[j] = temp
                    }
                }
            }

            for (let i = 0; i < StandardMultiboard.PlayersList.length; i++) {
                const player = StandardMultiboard.PlayersList[i]
                const currentStats = Globals.ALL_KITTIES.get(player)!.CurrentStats
                const playerColor = ColorUtils.GetStringColorOfPlayer(player.id + 1)

                const name = player.name.length > 8 ? player.name.substring(0, 8) : MapPlayer.name
                const score = currentStats.TotalSaves - currentStats.TotalDeaths
                const kda =
                    currentStats.TotalDeaths === 0
                        ? currentStats.TotalSaves.toFixed(2)
                        : (currentStats.TotalSaves / currentStats.TotalDeaths).toFixed(2)

                StandardMultiboard.PlayerStats[0] = name
                StandardMultiboard.PlayerStats[1] = score.toString()
                StandardMultiboard.PlayerStats[2] = currentStats.TotalSaves.toString()
                StandardMultiboard.PlayerStats[3] = currentStats.TotalDeaths.toString()
                StandardMultiboard.PlayerStats[4] = currentStats.SaveStreak.toString()
                StandardMultiboard.PlayerStats[5] = kda
                StandardMultiboard.PlayerStats[6] = `${currentStats.RoundSaves} / ${currentStats.RoundDeaths}`

                for (
                    let j = 0;
                    j < StandardMultiboard.PlayerStats.length - 1;
                    j++ // skip last element
                ) {
                    StandardMultiboard.CurrentStats.GetItem(rowIndex, j).setText(
                        `${playerColor}${StandardMultiboard.PlayerStats[j]}${Colors.COLOR_RESET}`
                    )
                    if (j === 0) StandardMultiboard.CurrentStats.GetItem(rowIndex, j).setWidth(0.07)
                }

                rowIndex++
            }
        } catch (ex: any) {
            print(`${Colors.COLOR_DARK_RED}Error in CurrentGameStats multiboard: ${ex}`)
        }
    }

    private static OverallGameStats = () => {
        StandardMultiboard.OverallStats.title = `Overall Stats ${Colors.COLOR_YELLOW_ORANGE}[${CurrentGameMode.active}-${Difficulty.DifficultyOption.toString()}]|r ${Colors.COLOR_RED}[Press ESC]|r`
        StandardMultiboard.OverallStats.rows = Globals.ALL_PLAYERS.length + 1
        let rowIndex = 1

        StandardMultiboard.PlayersList.length = 0
        for (let i = 0; i < Globals.ALL_PLAYERS.length; i++) {
            StandardMultiboard.PlayersList.push(Globals.ALL_PLAYERS[i])
        }

        // Sort the array of keys based on custom criteria
        for (let i = 0; i < StandardMultiboard.PlayersList.length; i++) {
            for (let j: number = i + 1; j < StandardMultiboard.PlayersList.length; j++) {
                const stats1 = Globals.ALL_KITTIES.get(StandardMultiboard.PlayersList[i])!.SaveData.GameStats
                const stats2 = Globals.ALL_KITTIES.get(StandardMultiboard.PlayersList[j])!.SaveData.GameStats
                const score1 = stats1.Saves - stats1.Deaths
                const score2 = stats2.Saves - stats2.Deaths

                if (
                    score2 > score1 ||
                    (score2 === score1 && StandardMultiboard.PlayersList[j].id < StandardMultiboard.PlayersList[i].id)
                ) {
                    const temp = StandardMultiboard.PlayersList[i]
                    StandardMultiboard.PlayersList[i] = StandardMultiboard.PlayersList[j]
                    StandardMultiboard.PlayersList[j] = temp
                }
            }
        }

        for (let i = 0; i < StandardMultiboard.PlayersList.length; i++) {
            const player = StandardMultiboard.PlayersList[i]
            const saveData = Globals.ALL_KITTIES.get(player)!.SaveData
            const playerColor = ColorUtils.GetStringColorOfPlayer(player.id + 1)

            const name = player.name.length > 8 ? player.name.substring(0, 8) : MapPlayer.name
            const allSaves = saveData.GameStats.Saves
            const allDeaths = saveData.GameStats.Deaths
            const score = allSaves - allDeaths
            const kda = allDeaths === 0 ? allSaves.toFixed(2) : (allSaves / allDeaths).toFixed(2)
            const gameCount = StandardMultiboard.GetGameCount(saveData)
            const winCount = StandardMultiboard.GetWinCount(saveData)

            StandardMultiboard.PlayerStats[0] = name
            StandardMultiboard.PlayerStats[1] = score.toString()
            StandardMultiboard.PlayerStats[2] = allSaves.toString()
            StandardMultiboard.PlayerStats[3] = allDeaths.toString()
            StandardMultiboard.PlayerStats[4] = saveData.GameStats.HighestSaveStreak.toString()
            StandardMultiboard.PlayerStats[5] = kda
            StandardMultiboard.PlayerStats[6] = gameCount.toString()
            StandardMultiboard.PlayerStats[7] = winCount.toString()

            for (let j = 0; j < StandardMultiboard.PlayerStats.length; j++) {
                StandardMultiboard.OverallStats.GetItem(rowIndex, j).setText(
                    `${playerColor}${StandardMultiboard.PlayerStats[j]}${Colors.COLOR_RESET}`
                )
                if (j === 0) StandardMultiboard.OverallStats.GetItem(rowIndex, j).setWidth(0.07)
            }

            rowIndex++
        }
    }

    private static BestTimesStats = () => {
        StandardMultiboard.BestTimes.title = `Best Times ${Colors.COLOR_YELLOW_ORANGE}[${CurrentGameMode.active}-${Difficulty.DifficultyOption.toString()}]|r ${Colors.COLOR_RED}[Press ESC]|r`
        StandardMultiboard.BestTimes.rows = Globals.ALL_PLAYERS.length + 1
        let rowIndex = 1

        for (let i = 0; i < Globals.ALL_PLAYERS.length; i++) {
            const player = Globals.ALL_PLAYERS[i]
            const saveData = Globals.ALL_KITTIES.get(player)!.SaveData
            const playerColor = ColorUtils.GetStringColorOfPlayer(player.id + 1)

            const roundTimes = StandardMultiboard.GetGameRoundTime(saveData)

            for (let j = 0; j < roundTimes.length; j++) {
                if (roundTimes[j] !== 0)
                    StandardMultiboard.BestTimes.GetItem(rowIndex, j + 1).setText(
                        `${playerColor}${Utility.ConvertFloatToTime(roundTimes[j])}${Colors.COLOR_RESET}`
                    )
                else
                    StandardMultiboard.BestTimes.GetItem(rowIndex, j + 1).setText(
                        `${playerColor}---${Colors.COLOR_RESET}`
                    )
            }
            rowIndex++
        }
    }

    public static UpdateOverallStatsMB = () => {
        if (CurrentGameMode.active !== GameMode.Standard) return
        StandardMultiboard.OverallGameStats()
    }

    public static UpdateStandardCurrentStatsMB = () => {
        if (CurrentGameMode.active !== GameMode.Standard) return
        StandardMultiboard.CurrentGameStats()
    }

    public static UpdateBestTimesMB = () => {
        if (CurrentGameMode.active !== GameMode.Standard) return
        StandardMultiboard.FillPlayers(StandardMultiboard.BestTimes, 1)
        StandardMultiboard.BestTimesStats()
    }

    public static FillPlayers(mb: Multiboard, rowIndex = 2) {
        for (const player of Globals.ALL_PLAYERS) {
            const name = player.name.length > 8 ? player.name.substring(0, 8) : MapPlayer.name
            mb.GetItem(rowIndex, 0).setText(`${ColorUtils.GetStringColorOfPlayer(player.id + 1)}${name}|r`)
            mb.GetItem(rowIndex, 0).setWidth(0.07)
            rowIndex++
        }
    }

    private static GetGameCount(data: KittyData) {
        const gameData = data.GameStats
        switch (Difficulty.DifficultyValue) {
            case DifficultyLevel.Normal:
                return gameData.NormalGames
            case DifficultyLevel.Hard:
                return gameData.HardGames
            case DifficultyLevel.Impossible:
                return gameData.ImpossibleGames
            case DifficultyLevel.Nightmare:
                return gameData.NightmareGames
            default:
                print(`${Colors.COLOR_DARK_RED}Error getting game count.`)
                return 0
        }
    }

    private static GetWinCount(data: KittyData) {
        const gameData = data.GameStats
        switch (Difficulty.DifficultyValue) {
            case DifficultyLevel.Normal:
                return gameData.NormalWins
            case DifficultyLevel.Hard:
                return gameData.HardWins
            case DifficultyLevel.Impossible:
                return gameData.ImpossibleWins
            case DifficultyLevel.Nightmare:
                return gameData.NightmareWins
            default:
                print(`${Colors.COLOR_DARK_RED}Error getting win count.`)
                return 0
        }
    }

    private static GetGameRoundTime(data: KittyData): number[] {
        const gameData = data.RoundTimes

        switch (Difficulty.DifficultyValue) {
            case DifficultyLevel.Normal:
                StandardMultiboard.RoundTimes[0] = gameData.RoundOneNormal
                StandardMultiboard.RoundTimes[1] = gameData.RoundTwoNormal
                StandardMultiboard.RoundTimes[2] = gameData.RoundThreeNormal
                StandardMultiboard.RoundTimes[3] = gameData.RoundFourNormal
                StandardMultiboard.RoundTimes[4] = gameData.RoundFiveNormal
                break

            case DifficultyLevel.Hard:
                StandardMultiboard.RoundTimes[0] = gameData.RoundOneHard
                StandardMultiboard.RoundTimes[1] = gameData.RoundTwoHard
                StandardMultiboard.RoundTimes[2] = gameData.RoundThreeHard
                StandardMultiboard.RoundTimes[3] = gameData.RoundFourHard
                StandardMultiboard.RoundTimes[4] = gameData.RoundFiveHard
                break

            case DifficultyLevel.Impossible:
                StandardMultiboard.RoundTimes[0] = gameData.RoundOneImpossible
                StandardMultiboard.RoundTimes[1] = gameData.RoundTwoImpossible
                StandardMultiboard.RoundTimes[2] = gameData.RoundThreeImpossible
                StandardMultiboard.RoundTimes[3] = gameData.RoundFourImpossible
                StandardMultiboard.RoundTimes[4] = gameData.RoundFiveImpossible
                break
            case DifficultyLevel.Nightmare:
                StandardMultiboard.RoundTimes[0] = gameData.RoundOneNightmare
                StandardMultiboard.RoundTimes[1] = gameData.RoundTwoNightmare
                StandardMultiboard.RoundTimes[2] = gameData.RoundThreeNightmare
                StandardMultiboard.RoundTimes[3] = gameData.RoundFourNightmare
                StandardMultiboard.RoundTimes[4] = gameData.RoundFiveNightmare
                break
            default:
                print(`${Colors.COLOR_DARK_RED}Error getting gamestat data for multiboard.`)
                return StandardMultiboard.RoundTimes
        }
        return StandardMultiboard.RoundTimes
    }

    private static ESCPressed = () => {
        if (CurrentGameMode.active !== GameMode.Standard) return
        if (!getTriggerPlayer().isLocal()) return
        if (StandardMultiboard.CurrentStats.displayed) {
            StandardMultiboard.CurrentStats.display(false)
            StandardMultiboard.OverallStats.display(true)
        } else if (StandardMultiboard.OverallStats.displayed) {
            StandardMultiboard.OverallStats.display(false)
            StandardMultiboard.BestTimes.display(true)
        } else if (StandardMultiboard.BestTimes.displayed) {
            StandardMultiboard.BestTimes.display(false)
            StandardMultiboard.CurrentStats.display(true)
        }
    }
}
